src/fdosecrets/objects/Prompt.cpp

```diff
@@ -196,6 +206,54 @@ namespace FdoSecrets
             }
         }
 
+        if (m_conn && m_conn->connected() && !m_items.empty()) {
+            QString q = tr("An application requests access to the following entries:\n\n");
+            for (const auto& i : m_items) {
+                if (i) {
+                    q += QStringLiteral("%1 %2\n").arg(QChar(0x2022)).arg(i->backend()->title());
+                }
+            }
+            q += tr("\nApplication: %1\n\n").arg(m_conn->peerName());
+            q += tr("Do you allow access to these entries only, all entries or deny the access?");
+
+            auto* checkbox = new QCheckBox(tr("Always request confirmations when applications access secrets"));
+            checkbox->setChecked(true);
+            checkbox->setToolTip(tr("You can change this setting from \"Settings\" -> \"Secret-Service-Integration\" "
+                                    "-> \"Confirm when passwords are retrieved by clients\" too."));
+            QObject::connect(checkbox, &QCheckBox::stateChanged, [&](int state) {
+                FdoSecrets::settings()->setConfirmAccessItem(static_cast<Qt::CheckState>(state)
+                                                             == Qt::CheckState::Checked);
+            });
+            auto dialogResult = MessageBox::question(nullptr,
+                                                     tr("KeePassXC: Secret-service access requested"),
+                                                     q,
+                                                     MessageBox::Yes | MessageBox::YesToAll | MessageBox::No,
+                                                     MessageBox::NoButton,
+                                                     MessageBox::Raise,
+                                                     checkbox);
+
+            if (m_conn && m_conn->connected()) {
+                if (dialogResult == MessageBox::Yes) {
+                    for (const auto& i : m_items) {
+                        if (i) {
+                            m_conn->setAuthorizedItem(i->backend()->uuid());
+                            m_unlocked << i->objectPath();
+                        }
+                    }
+                } else if (dialogResult == MessageBox::YesToAll) {
+                    m_conn->setAuthorizedAll();
+                    for (const auto& i : m_items) {
+                        if (i) {
+                            m_unlocked << i->objectPath();
+                        }
+                    }
+                }
+            }
+        }
+
+        // if we've get all
+        emit completed(m_unlocked.isEmpty(), QVariant::fromValue(m_unlocked));
```

Comment:
This introduces a race condition between unlocking collection dialog and emitting the completed signal. The `doUnlock` call is async, and only raises the unlock database dialog without waiting for it to finish. Consider the following case:

we want to unlock collection A, and item 1, 2, 3 from collection B:

1. `c->doUnlock` raises unlock dialog for collection A
2. code continues and raises dialog for unlocking items and blocks here
3. user interact with the item unlock dialog
4. `prompt` method continues from step 2 and emits completed signal
5. the client application gets completed signal and proceed assuming the collection A is unlocked
6. the collection A is not unlocked yet and triggers assertion in the client application

This can be triggered by using secret-tool to store a password into a locked database: `secret-tool store --label='Test4' username testuser password a`

This is the reason why the completed signal was emitted from `UnlockCollectionsPrompt::collectionUnlockFinished`, which will be emitted by the unlock database dialog.

Another concern is that in terms of UX, it is also less ideal IMHO to show multiple dialogs at the same time. (It's not that common for a client to have both collection unlocking and item unlocking happen at the same time, but it's still possible).

Probably the unlocking of items can be moved into the `UnlockCollectionsPrompt::collectionUnlockFinished`, and only do that after all requested collections are unlocked or rejected. I.e., in the same place as where `emit completed` was.